### PR TITLE
feat(client): add support for pinning explicit versions in main commands

### PR DESCRIPTION
**Pull request by dmmordvi**
## Summary

Add an explicit-VERSION mode to the main `trdl` client commands (`use`, `exec`, `update`, `bin-path`, `dir-path`) so users can pin a raw release (`v1.2.3`) or a semver constraint (`>=1.2.0`, `1.2.*`, `~1.2`) alongside the existing GROUP/CHANNEL selection. Add matching server-side validation to keep the two selection modes unambiguous.

## Key changes

- `client/cmd/trdl/common.go`: introduce `isVersionArg` (strict `v`+digit prefix OR valid semver constraint) and `validateVersionArgs` to dispatch between GROUP/CHANNEL and VERSION modes.
- `client/cmd/trdl/{use,exec,update,bin_path,dir_path}.go`: extend each command with a VERSION branch; update usage lines to `REPO GROUP [CHANNEL] | REPO VERSION`.
- `client/pkg/client/client.go` + `interface.go`: new API — `UpdateRepoToVersion`, `UseRepoReleaseBinDir`, `ExecRepoReleaseBin`, `GetRepoReleaseBinDir`, `GetRepoReleaseDir`.
- `client/pkg/repo/update.go`: `UpdateToVersion` resolves a constraint to the greatest matching TUF release and marks it via `metafile/versions/<release>`.
- `client/pkg/repo/client.go`: `FindLocalReleaseByVersion` re-resolves a constraint against local pin markers; shared release-dir resolver refactored out of the channel path.
- `client/pkg/repo/clean_releases.go`: preserve pinned releases across `--autoclean`.
- `client/pkg/repo/use.go`: generate one source script per constraint (16-hex sha256 slug); emit env exports via `formatSourceScriptEnvExports` with per-shell literal/expression/unset modes; add `TRDL_USE_<REPO>_VERSION` (resolved release, evaluated at each shell source) and `TRDL_USE_<REPO>_VERSION_CONSTRAINT` (original selector).
- Symmetric telemetry env cleanup: source scripts and `Exec*Bin` clear the opposite mode's variables so a shell/binary never sees both sets.
- `server/path_publish.go`: reject group names starting with `"v"` in `ValidatePublishConfig` — a v-prefixed group would be indistinguishable from a version on the client side.
- `client/cmd/trdl/command/template.go`: fix the broken `{{if gt .Aliases 0}}` (compared `[]string` to `int`); enables the `Aliases:` help section and the new `download` alias on `update`.
- `client/cmd/trdl/command/md_docs.go`: render `## Aliases` in generated CLI partials; regenerate `docs/_includes/reference/cli/trdl_*.md`.
- Tests: `client/cmd/trdl/{common_test,version_flag_test}.go`, `client/pkg/repo/use_test.go` (env-export escaping + slug determinism), `e2e/tests/client/version_test.go` (exact/range/tilde/caret pins across `update`, `bin-path`, `dir-path`, `exec`, autoclean).
- CI: new `unit_client` job runs `task client:test:unit`; wired into the notify stage.

## Why

Users currently have to change publish configuration to expose a specific release version — there is no client-side path to pin one build reproducibly (e.g. in CI, Dockerfiles, or long-lived shell sessions). Pinning by exact version or by semver constraint gives users the same guarantees `go.mod`/`pip`/`npm` provide, without server-side coordination. The `download` alias on `update` reflects the second common use case (fetch-without-activate).

## Review focus / risks

- **Breaking: server-side v-prefixed group rejection.** Before this PR `semver.NewVersion("v3")` succeeded, so a `groups: [{name: v3}]` config was valid. It is now rejected at `publish` time. The client mode dispatcher relies on this invariant. Existing publish configs with v-prefixed group names must be renamed. Release notes / migration callout welcome.
- **Pinned releases accumulate.** `metafile/versions/<release>` markers are never expired and there is no `unpin` command. Repeated `trdl update REPO <constraint>` calls (e.g. daily in CI) pin every historical resolution result; `--autoclean` intentionally skips them. Consider a follow-up (`trdl unpin` or a TTL).
- **Shell startup cost shift.** Version-mode source scripts inline `$(trdl dir-path REPO 'constraint' | awk ...)` (`$(((trdl dir-path REPO 'constraint') -split '[\\/]')[-2])` on pwsh). Every new sourced shell forks `trdl` and acquires a shared release lock. Channel mode used a fixed literal.
- **Escaping surface.** `formatSourceScriptEnvExports` handles `'` in literal values (`'\''` unix, `''` pwsh) and keeps expressions in double quotes. The version constraint itself is inlined into the emitted script single-quoted without escaping — safe today because upstream `semver.NewConstraint` rejects `'`, but no defense in depth on the write side.
- **`isVersionArg` classification.** `^v\d`-prefixed → VERSION; parseable semver → GROUP; valid constraint → VERSION; otherwise → GROUP. Covered by `common_test.go`; worth a second pair of eyes on the ordering (semver-before-constraint is intentional so plain `1.2` stays a group).
- **`slugifyConstraint`.** Raw sha256 → 16 hex chars. `"1.2.0"` and `"=1.2.0"` produce different slugs → duplicate script files. Storage bloat only; no cleanup of the `scripts/` dir.
- **Template fix side-effect.** Cobra help output now renders `Aliases:` for the first time — any downstream doc scraper of `--help` will see new content.

### CODE REVIEW

File: client/cmd/trdl/common.go [hunk 1]
```diff
@@ -21,6 +23,37 @@ func ValidateChannel(channel string) error {
 	return nil
 }
 
+// isVersionArg reports whether the given positional argument is an explicit
```
Comment (alexey-igrychev) [inline client/cmd/trdl/common.go:28]: `isVersionArg` classifies by a first-character regex `^[<>=~^v]`, then feeds the arg to `semver.NewConstraint`. This misclassifies in both directions:

- **False negatives** — legitimate semver constraints that don't start with those characters silently fall through to group/channel parsing: `!=1.2.3`, `1.2.x`, `1.2.*`, hyphen ranges like `1.2 - 1.4`, or a plain unprefixed exact `1.2.3`.
- **False positives** — any pre-existing GROUP whose name starts with `v` (e.g. `vendor`, `v3`) or one of `<>=~^` is routed into version mode and then rejected by `semver.NewConstraint`. This is a silent backward-compat break for any published repo that uses such a group name.

Recommend either an explicit `--version <constraint>` flag (unambiguous), or attempting `semver.NewConstraint` first and only falling back to group parsing when it fails AND the shape is not a plain version. In either case, add unit-test coverage for the boundary cases (`vendor`, `!=1.2.3`, `1.2.x`, plain `1.2.3`).

(generated by pi-pi)

File: client/pkg/repo/use.go [hunk 2]
```diff
@@ -129,8 +159,22 @@ export PATH="$trdl_repo_bin_path${PATH:+:${PATH}}"
...
 }
 
-func (c Client) prepareSourceScriptBasename(group, channel, shell string, opts UseSourceOptions) string {
-	basename := fmt.Sprintf("use_%s_%s_%s", group, channel, shell)
+func formatSourceScriptEnvExports(shell string, envs []sourceScriptEnv) string {
+	lines := make([]string, 0, len(envs))
+	for _, env := range envs {
+		switch shell {
+		case "pwsh":
+			lines = append(lines, fmt.Sprintf("[System.Environment]::SetEnvironmentVariable('%s',\"%s\",[System.EnvironmentVariableTarget]::Process);", env.Name, env.Value))
```
Comment (alexey-igrychev) [inline client/pkg/repo/use.go]: `formatSourceScriptEnvExports` now emits

```
[System.Environment]::SetEnvironmentVariable('NAME',"VALUE",…)
```

using **double quotes** around the value, applied unconditionally in both channel and version modes. The previous inline template used single quotes (`'value'`), which PowerShell treats as literal. This is required for the new version-extractor `$(…)` value, but for the channel-mode `TRDL_USE_<REPO>_GROUP_CHANNEL` export it is a behavior change: any `$`, backtick, or `$()` sequence appearing in a group/channel value is now expanded at source time.

Suggest quoting per-entry — single-quoted (with single-quote-escape) for literal values, double-quoted only for the extractor — or emit the extractor via a separate pwsh statement so channel-mode stays literal.

(generated by pi-pi)

File: client/pkg/repo/update.go [hunk 3]
```diff
@@ -237,6 +259,46 @@ func (c Client) filterTargets(prefix string) (data.TargetFiles, error) {
...
+		return "", fmt.Errorf("parse semver constraint %q: %w", version, err)
+	}
+
+	var latestValid *semver.Version
+	var latestValidName string
+	for name := range targets {
+		if !strings.HasPrefix(name, releasesDir+"/") {
+			continue
+		}
+
```
Comment (alexey-igrychev) [inline client/pkg/repo/update.go:279]: `findRelease` iterates every `releases/*` TUF target and does

```go
releaseVersion, err := semver.NewVersion(versionPart)
if err != nil {
    return "", fmt.Errorf("parse semver version %q: %w", name, err)
}
```

A single non-semver release name in the repository (`edge`, `nightly`, `main`, `2024-01-01`, …) aborts the entire search — even when another release satisfies the constraint. The server does not enforce semver-only release names (see `server/pkg/publisher/publisher.go`, which only strips a leading `v` for listing), so this fails-closed against any legitimately mixed repo.

Suggest `continue` (optionally with a debug log) on `semver.NewVersion` parse error, rather than returning an error.

(generated by pi-pi)

File: client/pkg/repo/client.go [hunk 4]
```diff
@@ -249,3 +308,45 @@ func (c Client) releaseMetafile(release string) util.Metafile {
...
+	}
+
+	var lastValid *semver.Version
+	var lastValidName string
+	for _, file := range matches {
+		releaseName := strings.TrimPrefix(file, filepath.Join(c.metafileDir, "versions")+string(os.PathSeparator))
+		releaseVersion, err := semver.NewVersion(releaseName)
+		if err != nil {
+			return "", fmt.Errorf("parse semver version %q: %w", releaseName, err)
+		}
```
Comment (alexey-igrychev) [inline client/pkg/repo/client.go:337]: `FindLocalReleaseByVersion` has the same fail-closed pattern as `findRelease`: it globs `metafileDir/versions/*` and calls `semver.NewVersion(releaseName)` on each entry; any single unparseable entry aborts every by-version local op (`exec`, `bin-path`, `dir-path`).

Even if `UpdateToVersion` is the only writer today, a stray file, an entry left over from a future format change, or a manually-created artefact will silently break the flow. `continue` on parse error instead.

(generated by pi-pi)

File: client/pkg/repo/use.go [hunk 5]
```diff
@@ -158,22 +202,64 @@ func (c Client) syncSourceScriptFile(group, channel, name string, data []byte) (
...
 
+func (c Client) prepareVersionExtractor(shell, version string) (string, error) {
+	trdlBinaryPath, err := trdl.GetTrdlBinaryPath()
+	if err != nil {
+		return "", err
+	}
+
+	switch shell {
+	case "pwsh":
+		return fmt.Sprintf("$(((%s dir-path %s '%s') -split '[\\\\/]')[-2])", trdlBinaryPath, c.repoName, version), nil
```
Comment (alexey-igrychev) [inline client/pkg/repo/use.go:250]: The pwsh branch of `prepareVersionExtractor` interpolates `trdlBinaryPath` with `%s`:

```go
fmt.Sprintf("$(((%s dir-path %s '%s') -split '[\\\\/]')[-2])", trdlBinaryPath, c.repoName, version)
```

When the binary lives under `C:\Program Files\…` PowerShell tokenizes the path as two arguments and the command breaks. The unix branch uses `%q`; the pwsh branch should too (with pwsh-appropriate escaping, e.g. `& '…\trdl.exe' …`). The pre-existing pwsh script template at the top of this file has the same issue; the new extractor extends it rather than fixing it.

(generated by pi-pi)

**Review by alexey-igrychev — COMMENTED** [review]
Review posted by pi-pi. 10 findings across the version-pinning flow. Highest-severity items:

- **BLOCKER** — mode detection by leading-character regex misclassifies in both directions (rejects legitimate groups starting with `v`/`<>=~^`, and silently mis-parses `!=`, wildcard, and plain-version constraints).
- **BLOCKER** — the pwsh env export switched from single- to double-quoted value, expanding `$`/backtick/`$()` in existing channel-mode group/channel values.
- **BLOCKER** — `findRelease` and `FindLocalReleaseByVersion` abort on the first non-semver release name; a single `edge`/`nightly`/dated release makes version mode unusable.
- **MAJOR** — pwsh `trdlBinaryPath` is embedded unquoted, breaking on `C:\Program Files\…`.
- **MAJOR** — six new methods on each of two exported `Interface`s is a source-breaking change for downstream implementers.

See the inline comments for the remaining minor items (misleading error message, `slugifyConstraint` misses `|`/`*`/`x`/`!`/`,`, docs contradiction, and a per-call `regexp.MustCompile`).

File: client/pkg/client/interface.go [hunk 6]
```diff
@@ -12,6 +12,11 @@ type Interface interface {
 	ExecRepoChannelReleaseBin(repoName, group, optionalChannel, optionalBinName string, args []string) error
 	GetRepoChannelReleaseDir(repoName, group, optionalChannel string) (string, error)
 	GetRepoChannelReleaseBinDir(repoName, group, optionalChannel string) (string, error)
+	UpdateRepoToVersion(repoName, version string, autocleanReleases bool) error
```
Comment (alexey-igrychev) [inline client/pkg/client/interface.go:15]: Six new methods are added to the exported `Interface`, and six more to `RepoInterface`. Since both are exported types, this is a source-breaking change for any downstream consumer that has its own mock or alternative implementation — they will fail to compile until they add the new methods.

Options: (1) split the version-mode surface into a separate optional interface, (2) drop the interface methods that no external caller needs and expose them only on the concrete `Client`/`RepoClient` types, or (3) document the break in the changelog.

(generated by pi-pi)

File: client/cmd/trdl/exec.go [hunk 7]
```diff
@@ -66,6 +78,32 @@ func processExecArgs(cmd *cobra.Command, args []string) (*execCmdData, error) {
...
+			restArgs = args[2:doubleDashInd]
+		}
+
+		switch len(restArgs) {
+		case 0:
+			return data, nil
+		case 1:
+			data.optionalBinaryName = restArgs[0]
+			return data, nil
+		default:
```
Comment (alexey-igrychev) [inline client/cmd/trdl/exec.go:100]: When more than one positional arg follows VERSION (`trdl exec repo v1.2.3 bin1 bin2`), the returned error is:

> VERSION is mutually exclusive with GROUP and CHANNEL arguments

But the user supplied neither GROUP nor CHANNEL — only two binary names. Suggest a message that describes the actual constraint, e.g. `"exec REPO VERSION accepts at most one BINARY_NAME"`.

(generated by pi-pi)

File: client/pkg/repo/use.go [hunk 8]
```diff
@@ -158,22 +202,64 @@ func (c Client) syncSourceScriptFile(group, channel, name string, data []byte) (
...
+}
+
 // formatRepoName returns a formatted repository name.
 // It replaces all non-alphanumeric characters with underscores and converts the result to uppercase.
 func formatRepoName(repoName string) string {
 	re := regexp.MustCompile("[^a-zA-Z0-9_]+")
 	formattedName := re.ReplaceAllString(repoName, "_")
 	return strings.ToUpper(formattedName)
 }
+
```
Comment (alexey-igrychev) [inline client/pkg/repo/use.go:280]: `slugifyConstraint` replaces `>=`, `<=`, `>`, `<`, `^`, `~`, `=`, and spaces, but does NOT cover `|` (compound `||`), `*`/`x` (wildcards), `!` (`!=`), or `,` (comma-joined ranges). `semver.NewConstraint` accepts constraints like `>=1.0 || <2.0` or `1.2.x`; these produce script/tmp directories containing `|` or `*`, which are illegal characters in Windows filenames.

Extend the replacer to cover these operators, or hash/base64 the constraint for the directory name to sidestep filesystem-character issues entirely.

(generated by pi-pi)

File: client/cmd/trdl/use.go [hunk 9]
```diff
@@ -17,39 +17,36 @@ func useCmd() *cobra.Command {
 	var shell string
 
 	cmd := &cobra.Command{
-		Use:   "use REPO GROUP [CHANNEL]",
+		Use:   "use REPO GROUP [CHANNEL] | REPO VERSION",
 		Short: "Generate a script to use the software binaries within a shell session",
 		Long:  `Generate a script to update the software binaries in the background and use local ones within a shell session`,
 		Example: `  # Source script in a shell
```
Comment (alexey-igrychev) [inline client/cmd/trdl/use.go:23]: The help example says:

```
# Pin an explicit version instead of a group/channel
# (an exact version must be prefixed with "v", otherwise it is treated as a group)
$ . $(trdl use repo_name v1.2.3)

# Pin a semver constraint (resolves to the greatest matching release)
$ . $(trdl use repo_name '>=1.2.0')
```

The first comment claims the version "must be prefixed with `v`", but the very next example uses `>=1.2.0` without a `v`. The actual rule is broader (any of `<>=~^v`) and doesn't match the docs. Reword the comment, or replace the rule with something the docs can state honestly (e.g. a `--version` flag).

(generated by pi-pi)

File: client/cmd/trdl/common.go [hunk 10]
```diff
@@ -21,6 +23,37 @@ func ValidateChannel(channel string) error {
...
+// release version. Versions are distinguished from a GROUP (a plain version
```
Comment (alexey-igrychev) [inline client/cmd/trdl/common.go]: `regexp.MustCompile("^[<>=~^v]")` is compiled on every call to `isVersionArg`, and `isVersionArg` is called at least once per invocation across five subcommands (and multiple times in `exec`). Hoist to a package-level `var versionArgRegexp = regexp.MustCompile(…)` to avoid the repeated compile.

(generated by pi-pi)
